client/src/main/scala/com.gu.contentapi.client/ContentApiClient.scala

```diff
@@ -106,8 +106,36 @@ trait ContentApiClient {
     paginate(query)(f).map {
       case Nil => throw new RuntimeException("Something went wrong with the query")
       case m :: Nil => m
-      case m1 :: m2 :: ms => ms.foldLeft(g(m1, m2))(g)
+      case ms => ms.reduce(g)
+    }
+
+  /** Unfolds a query by accumulating its results
+    * 
+    * @tparam Q the type of a Content API query with pagination parameters
+    * @tparam R the type of response expected for `Q`
+    * @param query the initial query
+    * @param f a result-processing function
+    * @return a future of an accumulated value
+    */
+  def paginateFold[Q <: PaginatedApiQuery[Q], R, M](query: Q)(f: (R, M) => M, m: M)(
+    implicit 
+    decoder: Decoder.Aux[Q, R],
+    decoderNext: Decoder.Aux[NextQuery[Q], R],
+    pager: PaginatedApiResponse[R],
+    context: ExecutionContext
+  ): Future[M] = {
+    def paginateFoldIn(nextQuery: Option[NextQuery[Q]])(m: M): Future[M] = {
+      val req = nextQuery.map(getResponse(_)).getOrElse(getResponse(query))
```

Comment:
Is this equivalent to the following?
```scala
val request = getResponse(nextQuery.getOrElse(query))
```

Reply:
Nope, because the two calls are different in the type of their respective implicit parameters. The compiler would compute the lub (which is `ContentApiQuery`) and that would break the constraints.